adding tests for externalclients and gcpservice

benzekrimaha opened the pull request:
Issue: ARSN-524

DarkIsDude commented on tests/unit/storage/data/DummyService.js:
```diff
@@ -88,6 +88,13 @@ class AzureDummyContainerClient {
     }
+
+    async deleteIfExists() {
+        if (this.key === 'externalBackendTestBucket/externalBackendMissingKey') {
+            return { succeeded: false };
+        }
+        return { succeeded: true };
+    }
```
I don't know how you imagine your futur but you can do :

`return { succeeded: this.key !== 'externalBackendTestBucket/externalBackendMissingKey' }`

DarkIsDude commented on tests/unit/storage/data/external/GcpService.spec.js:
```diff
@@ -108,7 +108,7 @@ const invalidDnsBucketNames = [
 function invalidDnsBucketNameHandler() {
     return (req, res) => {
-        assert(req.headers.host, host);
+         assert(req.headers.host, host);
```
Indent issue ?

DarkIsDude commented on tests/unit/storage/data/external/GcpService.spec.js:
```diff
@@ -329,3 +329,131 @@ describe('GcpService dnsStyle tests', () => {
+        });
+    });
+
+    it('createMultipartUpload should reject missing parameters', done => {
```
```suggestion
    it('createMultipartUpload should reject if parameters are missing', done => {
```

DarkIsDude replied:
same for next one

bert-e commented:
# Hello benzekrimaha,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/arsenal/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/ARSN-524` and the
destination branch `development/8.3`.

Please resolve the conflict on **the feature branch** (`improvement/ARSN-524`).

```sh
git fetch && \
git checkout origin/improvement/ARSN-524 && \
git merge origin/development/8.3
```

Resolve merge conflicts and commit

```sh
git push origin HEAD:improvement/ARSN-524
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

benzekrimaha committed:
rename method and file names for gcpApis

This commit aims to improve code clarity by renaming
methods and files related to GCP APIs
to more descriptive names.
The changes include renaming multipart upload methods
and object tagging methods to better reflect their
functionality.
Enhancing code readability and maintainability.
Issue: ARSN-524

benzekrimaha committed:
remove unused gcp client methods

This commit remove unused methods from the GcpClient class
as they are not needed anymore.
Issue: ARSN-524

benzekrimaha committed:
remove command wrapper function in gcpService

This commit tackles both the removal of the command wrapper functions in the GcpService
as they add no additional functonality, calling the send method directly is a better approach.
the class name was also updated to GcpService to better reflect its purpose and avoid
confusion with the client.
Issue: ARSN-524

benzekrimaha committed:
use aws commands instead of gcpService wrapper functions

Issue: ARSN-524

benzekrimaha committed:
coverage amelioration tests for externalclients

Some multipart upload tests were missing coverage for ExternalClients.
This commit adds those missing tests.
Issue: ARSN-524

benzekrimaha committed:
coverage amelioration tests for gcpService

With the refactor of GCP service within the migration from
aws-sdk v2 to v3, we noticed that some code paths were not covered by unit tests.
this commit adds tests to cover helper functions and mpu/object tagging related functions.
Issue: ARSN-524

francoisferrand commented on lib/storage/data/external/GCP/GcpService.js:
does not seem to be used either

francoisferrand commented on lib/storage/data/external/GCP/GcpService.js:
```diff
@@ -616,6 +569,6 @@ function gcpSigningMiddleware(config) {
 }
 
-Object.assign(GcpClient.prototype, gcpApis);
+Object.assign(GcpService.prototype, gcpApis);
```
can we move this to the constructor of GcpService, to make it easier to notice?

(or even rewrite this without `Object.assign` : it was done to "extend" the generated class, but now the class is manually written: so we may more cleanly add the methods?)

benzekrimaha replied:
Keeping GcpApis separate avoids a noisy/huge diff and makes it easier to iterate on these highly customized APIs in isolation. I kept the Object.assign(GcpService.prototype, gcpApis) at module scope (not in the constructor) because putting it in the constructor would re-run for every new GcpService(), repeatedly overwriting prototype methods.

SylvainSenechal replied:
Yeah not much we can do here. Maybe replace this line with something explicit
```suggestion
GcpClient.prototype.abortMultipartUpload = gcpApis.abortMultipartUpload;
GcpClient.prototype.completeMultipartUpload = gcpApis.completeMultipartUpload;
GcpClient.prototype.createMultipartUpload = gcpApis.createMultipartUpload;
...
...
...
```

This way it won't be 'hidden' into a single line, but not sure its super nice either

francoisferrand replied:
> putting it in the constructor would re-run for every new GcpService(), repeatedly overwriting prototype methods.

```
   constructor() {
     Object.assign(this, gcpApis);
   }
```

putting it in the constructor would not overwrite prototype method: only the "current" object's methods.
Would that be an issue?

francoisferrand commented on lib/storage/data/external/GCP/GcpService.js:
does not seem to be used
(please double check every single "legacy" method in here, everything we drop is something we don't have to maintain anymore)

francoisferrand commented on lib/storage/data/external/GCP/GcpService.js:
does not seem to be called?

francoisferrand commented on lib/storage/data/external/GCP/GcpService.js:
don't seem to be called?

francoisferrand commented on tests/unit/storage/data/external/GcpService.spec.js:
```diff
@@ -347,6 +364,435 @@ describe('GcpService dnsStyle tests', () => {
+    });
+});
+
+describe('GcpService mpu helper behavior', () => {
```
should be split in multiple tests: for each separate file (GcpService & each file in GcpApis)

francoisferrand commented on tests/unit/storage/data/external/GcpService.spec.js:
```diff
@@ -300,7 +317,7 @@ describe('GcpService pathStyle tests', () => {
 
     operations.forEach(test => it(`GCP::${test.op}`, done => {
-        client[test.op](test.params, err => {
+        callOperation(client, test.op, test.params, err => {
```
these test execute the code (i.e. give some code coverage), but do not verify that code behaves as expected: so not so useful... should add some assertions on the result or calls it makes...

francoisferrand commented on tests/unit/storage/data/external/GcpService.spec.js:
```diff
@@ -181,6 +184,20 @@ const operations = [
+            .catch(err => cb(err));
+    }
+    return client[op](params, cb);
+}
```
what is the point to introduce such a wrapper?
it really just means that listObejcts and listObject should not be handled through the same `operations` table : either as extra (separate) tests, or through another table which would contain the type object to create or object to send...

francoisferrand commented on tests/unit/storage/data/external/GcpService.spec.js:
```diff
@@ -347,6 +364,435 @@ describe('GcpService dnsStyle tests', () => {
+    });
+
+    it('putObjectTagging should merge tags into metadata', done => {
+        jest.spyOn(client, 'headObject')
```
we usually use `Sinon.js` for mocks : any reason to switch to `jest` ? or is it not the case in arsenal?

(nothing against jest mocks in itself, but consistency is a good thing... especially with untyped languages like JS)

benzekrimaha replied:
Switched to sinon

francoisferrand commented on tests/unit/storage/data/external/GcpService.spec.js:
```diff
@@ -347,6 +364,435 @@ describe('GcpService dnsStyle tests', () => {
+        jest.restoreAllMocks();
+    });
+
+    it('abortMultipartUpload should reject missing parameters', done => {
```
use `describe` to provide context, so that the text like looks like a proper sentence: "it should reject missing parameters"

```suggestion
describe('abortMultipartUpload', () => {
   it('should reject missing parameters', done => {
   
   [...]
})

describe('completeMultipartUpload', () => {
   [...]
```

(same for all these test)

francoisferrand commented on tests/unit/storage/data/external/GcpService.spec.js:
```diff
@@ -347,6 +364,435 @@ describe('GcpService dnsStyle tests', () => {
+    });
+
+    describe('removeParts', () => {
+        it('should list versions via send(ListObjectVersionsCommand) and delete each version', done => {
```
```suggestion
        it('should list object versions and delete them all', done => {
```

francoisferrand commented on tests/unit/storage/data/external/GcpService.spec.js:
```diff
@@ -347,6 +364,435 @@ describe('GcpService dnsStyle tests', () => {
+            });
+        });
+
+        it('should return error when send(ListObjectVersionsCommand) rejects', done => {
```
```suggestion
        it('should return error when list objects command fails', done => {
```

francoisferrand commented on tests/unit/storage/data/external/GcpService.spec.js:
```diff
@@ -347,6 +364,435 @@ describe('GcpService dnsStyle tests', () => {
+            });
+        });
+
+        it('should return error when command rejects', done => {
```
```suggestion
        it('should return error when command fails', done => {
```

benzekrimaha committed:
removing unused methods in gcpService and updating tests to use send command

Issue: ARSN-524

benzekrimaha committed:
gcpApis tests introduction

Issue: ARSN-524

benzekrimaha committed:
Remove gcpApis related tests and update existing GcpService tests

Issue: ARSN-524

SylvainSenechal commented on tests/unit/storage/data/DummyService.js:
```diff
@@ -88,6 +88,12 @@ class AzureDummyContainerClient {
         };
     }
+
+    async deleteIfExists() {
```
I don't see this function being used anywhere 🤔 except once in code that was already existing in azureClient test. How come the test didn't fail without this function

SylvainSenechal commented on tests/unit/storage/data/external/ExternalClients.spec.js:
```diff
@@ -167,6 +185,16 @@ describe('external backend clients', () => {
         });
 
+        it(`${backend.name} delete() should delete the requested key without error`, async () => {
+            const key = 'externalBackendTestKey';
```
nit: seems these key/bucket names are important as they are also hardcoded in the dummy service, but we could consider using constants dummyKey/dummyBucketName defined in dummy service and importing them

francoisferrand reviewed (approved):
with this, are we ready for 8.3.0 release?
- if we don't have anything, should bump in this PR so we can immediately release
- or align with @DarkIsDude to release once https://github.com/scality/Arsenal/pull/2581 is merged (on 8.2, with 8.3 waterflow), merge this one "as is", and do the bump #2581.

benzekrimaha commented:
/approve

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.3`



The following branches have **NOT** changed:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.1`
* `development/8.2`



Please check the status of the associated issue ARSN-524.

Goodbye benzekrimaha.




*The following options are set:* **approve**
